specs/_features/eip6800/beacon-chain.md

```diff
@@ -0,0 +1,221 @@
+    - [`IPAProof`](#ipaproof)
+    - [`VerkleProof`](#verkleproof)
+    - [`ExecutionWitness`](#executionwitness)
+- [Beacon chain state transition function](#beacon-chain-state-transition-function)
+  - [Block processing](#block-processing)
+    - [Execution payload](#execution-payload)
+      - [`process_execution_payload`](#process_execution_payload)
+- [Testing](#testing)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This upgrade adds transaction execution to the beacon chain as part of the eip6800 upgrade.
+
+## Custom types
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `BanderwagonGroupElement` | `Bytes32` | |
+| `BanderwagonFieldElement` | `Bytes32` | |
+| `Stem` | `Bytes31` | |
+
+## Preset
+
+### Execution
+
+| Name | Value |
+| - | - |
+| `MAX_STEMS` | `uint64(2**16)` (= 65,536) |
+| `MAX_COMMITMENTS_PER_STEM` | `uint64(33)` |
+| `VERKLE_WIDTH` | `uint64(2**8)` (= 256) |
+| `IPA_PROOF_DEPTH` | `uint64(2**3)` (= 8) |
+
+## Containers
+
+### Extended containers
+
+#### `ExecutionPayload`
+
+```python
+class ExecutionPayload(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress  # 'beneficiary' in the yellow paper
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32  # 'difficulty' in the yellow paper
+    block_number: uint64  # 'number' in the yellow paper
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions: List[Transaction, MAX_TRANSACTIONS_PER_PAYLOAD]
+    withdrawals: List[Withdrawal, MAX_WITHDRAWALS_PER_PAYLOAD]
+    blob_gas_used: uint64
+    excess_blob_gas: uint64
+    execution_witness: ExecutionWitness  # [New in EIP6800]
+```
+
+#### `ExecutionPayloadHeader`
+
+```python
+class ExecutionPayloadHeader(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32
+    block_number: uint64
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawals_root: Root
+    blob_gas_used: uint64
+    excess_data_gas: uint64
+    execution_witness_root: Root  # [New in EIP6800]
+```
+
+### New containers
+
+#### `SuffixStateDiff`
+
+```python
+class SuffixStateDiff(Container):
+    suffix: Bytes1
+    # Null means not currently present
+    current_value: Optional[Bytes32]
+    # Null means value not updated
+    new_value: Optional[Bytes32]
```

Comment:
These should either be updated for EIP-7495 `StableContainer` / `Profile` (on `SuffixStateDiff`), or unwrapped so that e.g. a zero value is used instead of `None`. Would like to withdraw EIP-6475 style `Optional` as the functionality is covered by the more advanced EIP-7495 as well.

SSZ library implementation progress here: https://stabilitynow.box

Reply:
Oh right.

@gballet 
The current `Optional` in pyspec is the [Python typing.Optional](https://docs.python.org/3/library/typing.html#typing.Optional). Our SSZ lib [Remerkleable](https://github.com/protolambda/remerkleable) only implemented SSZ `Union`, so technically we can use `Union[None, Bytes32]` here in the specs. However, other SSZ libs didn't implement `Union` and may go implement the newer SSZ types as @etan-status suggested.

Is it possible to use `Bytes32()`, the empty `0x00...00`, to represent null in this case?

Reply:
no it is not possible to use `Bytes32()` and I was initially using `Union[None, Byte32]` and was asked to replace it with `Optional``. Precisely because `Union` didn't exist (except in the wonderful gballet/ssz.zig, of course :innocent:).